# Introduction to phylogenetics {#phylogenetics}

Most of the topics in this book are just introduced, and when you start using the tools you will
have to study the original literature. This is true to a probably greater extent for phylogenetics
which is a truly vast area of computational methodology. It is also one which has for long been
fraught by strongly held opinions bordering on methodological wars. In this chapter I will try to
give an overview of available tools and methods, together with some advice on how to address typical
situations and problems. I will try to be neutral, but at times my own opinions might shine through,
since I actually want to convey some personal expericences of mine in the hope they might be useful
to you.

## \index{multiple sequence alignment}Multiple sequence \index{alignment}alignments, homology and trees {#multiple-sequence-alignment}

Today, the most common starting point to estimate a phylogenetic tree are nucleotide or protein
sequences. Moreover, the sequences are aligned so that corresponding positions are placed at the
same relative positions. The idea behind aligning sequences like this is that we can capture
nucleotides or amino acids that are \index{homology}*homologous*, i.e. are derived from the same
nucleotide or amino acid state in a common ancestral sequence. This is similar to the alignments
that were discussed in the chapter about annotation (chapter \@ref(annotation)), but whereas they
were *pairwise*, the alignments you use for estimating trees are *multiple*, i.e. contain more than
two sequences.

An alignment of a few rather similar protein sequences can look like this (in *phylip* format):

```
9   49
S._cerevisiae_NrdAe             GLIPMIRVFN NTARYVDQGG NKRPGAFALF LEPWHADIFD FVDIRKTHG
E._coli_KTE27_NrdAg             GCIPFYKHFQ TAVKSCSQGG -VRGGAATLF YPMWHLEVES LLVLKNNRG
Clostridium_acetobutylicum_ATC  GVVPWVKLFN DTAIAVDQLG -VRNGSVAVW LDAWHKDLPE FMQLRTNNG
T._denticola_NrdAh              GIIRWIKLAN DTAVAVDQLG -VRQGSVAVY LDVWHKDIPE FLQLRTNNG
A._baumannii_NrdAi              GAREVIDDFF TAASKISQGG -NRRGSFAAY LDIEHDDFYE CLDSLRTHS
Streptomyces_sp._NrdAn          GIVPFLKTLD ASVAAVNQGG -RRKGAAAVY LETWHADIEE FLELRDNTG
Delftia_sp._NrdAz               GVVPFLKVVN DTAVAVNQGG -KRKGAVCTY LETWHLDIEE FLELRKNTG
S._aureus_NrdE                  GVLPIAKSLE GGFSYADQLG -QRPGAGAVY LNIFHYDVEE FLDTKKVNA
B._subtilis_NrdE                GVVGVMKLLD NAFRYADQMG -QRQGSGAVY LSVFHPDITD FLDTKKISA
```

It is often better to look at sequence alignments in specialized programs, e.g.
[Seaview](http://doua.prabi.fr/software/seaview) [@gouy_seaview_2010],
[ClustalX](http://www.clustal.org/) [@thompson_multiple_2003] or [JalView](http://www.jalview.org/)
[@clamp_jalview_2004] (avoid the alignment program, ClustalW, in ClustalX; there are better
alternatives today).

![*View of a multiple sequence alignment in Seaview*. Programs that display and lets you work with
multiple sequence alignments colours amino acids so it's easier to see *conservative substitutions*,
i.e. substitutions between biochemically similar amino acids.](img/seaview_screenshot.png)

When you run a phylogenetic tree building software it will consider each column as a separate
character with different *states* in each sequence. This information will then be used to estimate
the phylogeny.

## \index{phylogenetic algorithms}Different types of tree building algorithms {#phylogenetic-algorithms}

There are essentially three types of algorithms used to estimate a phylogeny from a set of aligned
sequences:

1. Distance-based

   Distance-based methods calculate a pairwise distance between all sequences in the alignment, and
   then uses the distances to group sequences separated by short distances nearby each other in the
   tree. The most widely used algorithm in this class is \index{neighbor joining}*neighbor joining*
   -- affectionately known as *NJ* -- which is a fast \index{heuristic algorithm}*heuristic*,
   \index{greedy algorithm}*greedy* algorithm (see appendix \@ref(computer-terms), Computer terms)
   [@saitou_neighbor-joining_1987].  There are also non-greedy algorithms that aims to find the
   shortest tree, i.e. the tree with the shortest possible sum of branch lengths where branch
   lengths are proportional to distances. An example of a non-greedy algorithm is FastME (ME is
   short for *Minimum Evolution*) [@lefort_fastme_2015].

2. Parsimony

   \index{parsimony}*Parsimony* in phylogenetics builds on the famous philosophical principle known
   as *Occam's razor* which contends that "simpler solutions are more likely to be correct than
   complex ones" ([quote from Wikipedia](https://en.wikipedia.org/wiki/Occam%27s_razor)). This is
   interpreted in phylogenetics as "find the tree that minimizes the number of changes of state".
   States were, before sequences were available, morphological character states. A tree of
   vertebrates would hence suggest three origins of powered flight (birds, bats and pterosaurs) --
   three transitions of state in the forelimbs from legs to wings --
   considering presence or absence of wings, birth of live offspring, the shape of the pelvis and a
   few other characters. In sequence based parsimony, state transitions are transitions between the
   different nucleotides or amino acids. 

3. Probabilistic
   
   Probabilistic phylogenetic methods are based on a statistical model of sequence evolution. They
   come in two important "flavours": \index{maximum likelihood phylogenetics}*maximum likelihood*
   (ML) and \index{Bayesian phylogenetics}Bayesian. In simplified terms, maximum likelihood
   algorithms tries to find the tree that makes the input data, i.e. the alignment, most probable
   given all other model parameters. Bayesian methods instead searches for the phylogeny that has
   the highest \index{posterior probability}*posterior probability* given the data and parameters.

   The current ML algorithms (e.g. PhyML [@guindon_estimating_2009], RAxML [@stamatakis_raxml_2014]
   and FastTree [@price_fasttree_2010]) are many times faster, and can hence deal with much larger
   alignments than Bayesian algorithms (e.g. MrBayes [@ronquist_mrbayes_2003], BEAST
   [@bouckaert_beast_2019] and PhyloBayes [@lartillot_phylobayes_2009]). There are, however, special
   applications which Bayesian methods are better suited to (e.g. moleculcar clocks and parameter
   rich models) and some would argue that they also better deal with problematic data.

### Exploring tree space

The parsimony and probabilistic phylogenetic algorithms look for the optimal tree, which requires
that all possible trees are checked. Since this is a very large number for any reasonable number of
sequences (there are, e.g., $10^{76}$ possible rooted trees with 50 species/sequences), any
algorithm will hence have to be \index{heuristic algorithm}heuristic. They are, as far as I know,
never *greedy* like the most common distance based methods. Instead, parsimony algorithms rely on
different ways of *exploring tree space* and calculating parsimony scores on each. If you're
interested in how this is done, I can suggest e.g. Joseph Felsenstein's seminal book *Inferring
phylogenies* @felsenstein_inferring_2004.

## A workflow for protein phylogenetics {#protein-phylogenetics-wf}

The above aims to provide a gentle introduction to concepts and tools in phylogenetics. This section
instead aims to outline a workflow for estimation of phylogenies for individual proteins. The
typical case where you would run this is when you want to find out which sequences are orthologs and
paralogs in a set of likely homologs. As described in the section about orthology and paralogy
(\@ref(orthology-paralogy)), this can be very important as support for claiming that a protein found
in your organism is doing the same as a specific protein in another organism. This is not the same
as finding the most similar sequence. Instead, it is relatively common that similarity does not
confer the signal of evolutionary relatedness that's required to establish orthology.

1. Collect sequences

   To establish evolutionary relationships among members of protein families, it is critical to
   collect sequences that well represent the family. If branches of the real evolutionary tree are
   not represented with sequences, your tree will likely be misleading. The key tools here are
   profile search tools like HMMER [@eddy_accelerated_2011] (see \@ref(profile-search-tools)) as
   they are the most sensitive and precise tools we have for alignment searches.

   a. Find an HMM profile

   b. Build and HMM profile

   c. Refine your HMM profile

2. Cluster sequences

   For many protein families there are too many sequences available for aligning, estimating a
   phylogeny and analysing. It can hence be important to reduce the number of sequences, but doing
   this in a way that keeps the sequence collection representative for the protein family. I do this
   by clustering the sequences based on pairwise sequence similarity using USEARCH
   [@edgar_search_2010]. I run the program with a number of different similarity cutoffs until I get
   a reasonable number of sequences. What is "reasonable" depends on the situation -- a larger protein
   family will of course need more sequences than a smaller, longer proteins contain more
   information than shorter and would require more sequences, more divergent families will also
   require more sequences -- but a few hundred sequences are relatively easy to deal with, whereas
   more than 500 can be difficult to e.g. visualize.

3. Align sequences and delete problematic sequences

   When you have a number of sequences you belive well describe your protein family, an iterative
   work starts where you align all sequences, check the alignment, delete sequences that are
   suspicious by e.g. lacking certain parts, such as catalytic residues, that the other have or
   having long insertions^[ Deleting problematic sequences is a difficult task and you risk losing
   sequences that represent important evolutionary events. This is particularly true for sequences
   that are only found in parts of the tree of life that is underrepresented by sequenced genomes.
   Moreover, you risk only describing what you already knew and become blind for true and novel
   sequences.], and then realign. Also, keep an eye on differences in domain content between
   different members of the family. This is visible by relatively long insertions in one group of
   proteins compared to the other, and can be problematic^[If you're working with a protein family
   that contains multiple domains in a way you think might be problematic, find someone experienced
   to discuss with.]. You continue this process until you're sure you have a good selection of
   sequences.

   The tools I use for this part is the Clustal Omega alignment program
   [@sievers_clustal_omega_2014] (fast and good; don't use ClustalW), the Seaview alignment viewer
   [@gouy_seaview_2010] together with Vim, my text editor (see @ref(text-editors)).

4. Check "treelikeness" and conflicts in the aligned sequences

   Several evolutionary events can create sequences that are not well described by a
   \index{bifurcating tree}bifurcating tree^[A *bifurcating tree* is a tree where each internal node
   has one ingoing and two outgoing edges, usually representing speciation of an ancestral species
   into two derived species. This is likely the most common type of speciation, but not necessarily
   the only. Most (all?) programs that estimate phylogenies only produce bifurcating trees.]. If the
   tree describes species, the most common such events are \index{hybridization}hybridization and
   \index{horizontal gene transfer}horizontal gene transfer (HGT) that will create different
   evolutionary histories for different genes and proteins in a genome. On the scale of individual
   genes or proteins, the corresponding process is \index{recombination}recombination within a gene,
   by which the parts of a gene will come from different parent genes and you will hence have a gene
   or protein sequence with different evolutionary histories. Recombination leading to swapping of
   parts of genes between each other will not be visible in a bifurcating tree -- except by reduced
   support values, which can have many other reasons. 
   
   To describe these processes, \index{phylogenetic network}phylogenetic networks -- graphs where
   processes such as hybridization and horizontal gene transfer can be described for species and
   recombination for individual sequences -- are required. Tools that infer phylogenetic networks
   largely fall into two categories, those aiming to describe actual evolutionary events like
   hybridization and those that only aim to show conflicts in the data. The SplitsTree program
   [@huson_application_2006] can be used to study NeighborNet [@bryant_neighbor-net_2004]
   representations of alignments to find conflicts (as well as a number of other phylogenetic
   networks).

   ![A NeighborNet network showing conflicting signals in the data. The red and blue line partition
   the network in two contradictory "splits". Figure from
   [@morrison_introduction_2011].](img/morrison_2011_fig3.21.edit.png)

   After you have studied the network, it's quite common that you have to go back to removing
   sequences or even redoing the selection of starting sequences.

5. Finalize alignment

   When you're finally happy with the alignment, you can at this point choose to realign with an
   alignment program that you trust more, but may be slower. Personally, I like *Probcons*
   [@do_probcons_2005]. If your proteins are not very divergent, Clustal Omega is fine.

6. Select reliable positions in the alignment

   In most cases, not all positions of the alignment can be trusted, and there are algorithms that
   look for reliably aligned positions. I use the BMGE algorithm [@criscuolo_bmge_2010], which uses
   amino acid scoring matrices from the BLOSUM series (see \@ref(amino-acid-scoring-matrices)) to
   identify reliable positions.

7. Estimate phylogenetic tree

## \index{phylogenetic placement}Phylogenetic placement of many short sequences {#phylogenetic-placement}

In microbial ecology it's common that we only have fragments of genes, most commonly an amplicon
from a segment of a phylogenetic marker such as the 16S gene (see \@ref(amplicon-sequences)).
Building full phylogenies from fragments is usually not reliable and a better option is to place the
fragments in an already established reference phylogeny that was built from full length sequences.
Taking our cherished V3V4 16S amplicons as an example, it is better to find a phylogeny made from
known full length 16S sequences together with the alignmnent and place your amplicons in this. In
practice this is done by first aligning the fragments to the alignment used for the reference tree,
and then run a phylogenetic placement algorithm that takes both the new alignment and the reference
phylogeny as input. The most commonly used phylogenetic placement algorithm is probably
\index{pplacer}pplacer [@matsen_pplacer_2010] but also \index{RAxML}RAxML [@stamatakis_raxml_2014]
can be used.

QIIME2!
```{r include=FALSE}
# vim:tw=100
```
